### ISSUE TITLE
Comma is inserted between items and paths in macros

### ISSUE THREAD

**Issue by scileo** [rust-lang/rustfmt#4375]
TLDR: a `,` is inserted between items and other fragment specifiers in macros, which cause macros to break after formatting.

**Input**

```rust
macro_rules! my_macro {
    // We don't care about what the macro expands to.
    // What is important here is the rule.
    ($left:item $right:path) => {};
}

// Note: there is no colon between the ; and the std tokens
my_macro!(struct Foo; std::fs);
```

**Output**

For brevity, I have removed the macro definition.

```rust
my_macro!(struct Foo;, std::fs);
```

Notice the `,` inserted between `;` and `std`, which breaks earlier macro.

**Expected output**

```rust
my_macro!(struct Foo; std::fs);
```

**Note**

I have met this behavior when `$right` is a `path`, but this also happens if `path` is one of the following specifiers:
  - item,
  - block,
  - stmt,
  - pat,
  - expr,
  - ty,
  - ident,
  - path,
  - tt,
  - vis.

I have created a [playground link][link], so that it is easier to see what's wrong.

**Meta**

- rustfmt version: rustfmt 1.4.15-stable (530eadf 2020-06-02) (tested, also reproducible on the playground)
- From where did you install rustfmt?: Installed via NixOS package manager, but also happens on the playground.

Feel free to ping me if other information is needed.

[link]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=36220c87452fcc16b63b9a1216c91c56

**Comment by scileo** [comment]
> More sophisticated solutions would require us to look at how the macro is defined, which we don't have access to at the AST level. I am saying this because it's possible to fix, but would be kind-of hacky and may bring more problems.

Completely agree: this is not the best way to fix the issue. Additionally, this wouldn't solve the issue when the macro itself is not defined.

IMO no tokens should be added or removed in macros (unless the macro is defined for instance in the standard library), but I may be wrong.

**Comment by ytmimi** [comment]
Confirming this is still reproducible with `rustfmt 1.5.1-nightly (a451a39d 2022-07-25)`
